mat64/index_bound_checks.go

```diff
@@ -63,3 +65,47 @@ func (m *Vector) set(r int, v float64) {
 	}
 	m.mat.Data[r*m.mat.Inc] = v
 }
+
+// At returns the element at row r and column c.
+func (t *Symmetric) At(r, c int) float64 {
+	return t.at(r, c)
+}
+func (t *Symmetric) at(r, c int) float64 {
+	if r >= t.mat.N || r < 0 {
+		panic(ErrRowAccess)
+	}
+	if c >= t.mat.N || c < 0 {
+		panic(ErrColAccess)
+	}
+	if t.mat.Uplo == blas.Upper {
+		if r > c {
+			r, c = c, r
+		}
+		return t.mat.Data[r*t.mat.Stride+c]
+	}
+	if r < c {
+		r, c = c, r
+	}
+	return t.mat.Data[r*t.mat.Stride+c]
+}
+
+// Set sets the element at row r and column c. Set panics if the location is outside
+// the appropriate half of the matrix.
```

Comment:
I think it would be better for the matrix package as a whole if the matrix types tended to implement the same interfaces where possible.  In this case I think it should implement Set as @vladimir-ch suggests, and trust the users that they would access it correctly.  We can provide an additional SetSymer (or SymSetter)  that is only implemented by symmetric matrixes.  Then a type assertion could reveal to users that they only need to interact with half of a matrix, but if they aren't trying to be sophisticated they could still use Set and have it work correctly (even if it was inefficient).